### DISCUSSION TITLE
No subbranches showing up with TBranch.branches

### DISCUSSION THREAD

**Discussion by lillsantosolms** [scikit-hep/uproot5#419] [Q&A]
Hi, 

I am trying to get data from a root file using uproot4 and I am having trouble with some of the branches. When I use TBranch.branches, I get no subbranches even though there should be quite a few of them (I verified this using TBrowser in ROOT). Does anyone know what could be causing this issue? Also, is there any other way for me to access these subbranches using uproot? 

Thanks.

**Comment by jpivarski** [comment]
Well, `TBranch.branches` only goes one level deep, but `TBranch.keys()`, `TBranch.values()`, and `TBranch.items()` are recursive by default. But if you get _zero_ subbranches from `TBranch.branches`, then recursion won't help. Maybe also look in `TTree.show()`?

All of these methods go to the same place: the branches that are nested within the TTree/other TBranches (through a TObjArray). There's nowhere else to look.

**Reply by lillsantosolms** [reply]
Thanks for your response. I have tried TTree.show(), TBranch.keys(), TBranch.values(), and TBranch.items(). TTree.show() gave me this:

name                 | typename                                           | interpretation                                                                                      
---------------------+----------------------------------------------------+----------------------------------------------------------
Sim                  | VASimulationData                                   | AsStridedObjects(Model_VASimulationData_v8)

but no information about subbranches within 'sim'. And as you said, TBranch.keys(), TBranch.values(), and TBranch.items() gave me zero subbranches. I also tried accessing the data from this branch directly using TBranch.array(), but I got a ValueError saying that "basket 0 has the wrong number of bytes (13860) for interpretation AsStridedObjects(Model_VASimulationData_v8)". And when I tried Model_TBasket.array() on the other baskets of the branch, I pretty much got the same error. 

I thought about using AsStridedObjects.basket_array(), but I am not sure what to put for the 'cursor_offset' argument (I was able to find information about the other arguments in the documentation on TBranch objects and Model_TBasket objects). Am I on the right track or is there a better way to access the data from the 'sim' branch?

**Reply by jpivarski** [reply]
It looks like the TTree has exactly one TBranch, and that TBranch has zero subbranches. So `TTree.keys()` etc. should return a list of that one TBranch (and `TBranch.keys()` on the TBranch should show zero). They should be presenting a consistent picture—a strong "should" because they all access the same data.

ROOT's TBrowser may be presenting data members within that one TBranch as things to click on and get a plot, but they're not TBranches. Whether or not they're TBranches depends on whether the file was written with splitting on or off; ROOT considers that an implementation detail, while Uproot considers that user-visible. Since ROOT considers it an implementation detail, TBrowser would show you all the variables you can click on the same way, whether they're implemented in subbranches or a single TBranch. That could be why they look different.

Not being able to read that TBranch is a bug in Uproot; please report it as an Issue with an example file. Uproot is thinking `VASimulationData` a simple class type that can be read in a "strided" way (faster, avoids Python iteration, but only works if the class struct has fixed size, made of types like `int` and `float`, not strings or nested lists). Maybe Uproot is assigning the wrong interpretation and it needs to be a general object, or maybe something else is going wrong. To figure that out, I'll need to see an example of the file, so that I can `show_streamers("VASimulationData")` to verify whether it can be strided and see why the predicted size is not matching the actual TBasket size (which needs to be an integer multiple of a single `VASimulationData` object: that's what the error message is saying).

`basket_array` would give you the same error for the same reasons. There's also `debug` and `debug_array` methods that I'll probably use, which dump it as raw bytes (and therefore the size doesn't have to be an integer multiple of anything).
